File: src/y2020/day02.rs
use itertools::Itertools;

use crate::utils::str_idx;

type Parsed<'a> = &'a str;
pub fn parse<'a>(raw: &'a str) -> Vec<Parsed> {
    raw.split('\n').collect()
}

fn process(input: &Vec<Parsed>, f: fn((Vec<usize>, char, String)) -> bool) -> usize {
    let mut valid = 0;
    for x in input {
        let (num, c, pwd) = x.splitn(3, " ").collect_tuple().unwrap();
        let num = num
            .split("-")
            .map(|i| i.parse::<usize>().unwrap())
            .collect::<Vec<usize>>();

        let c = str_idx(&c.to_string(), 0);

        if f((num, c, pwd.to_string())) {
            valid += 1;
        }
    }
    valid
}

pub fn part1(input: &Vec<Parsed>) -> usize {
    let f = |(num, c, pwd): (Vec<usize>, char, String)| -> bool {
        (num[0]..=num[1]).contains(&(pwd.matches(c).count()))
    };
    process(input, f)
}

pub fn part2(input: &Vec<Parsed>) -> usize {
    let f = |(num, c, pwd): (Vec<usize>, char, String)| -> bool {
        (str_idx(&pwd, num[0] - 1) == c) ^ (str_idx(&pwd, num[1] - 1) == c)
    };
    process(input, f)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::utils::read;
    #[test]
    fn test1() {
        assert_eq!(part1(&parse(&read(2020, "day02.txt"))), 447);
    }

    #[test]
    fn test2() {
        assert_eq!(part2(&parse(&read(2020, "day02.txt"))), 249);
    }
}

File: python/y2020/day02.py
# %%
from collections import Counter

from utils import load

raw = load("day02.txt")

# %% Part 1
def test1() -> None:
    valid = 0
    for x in raw:
        num, char, pwd = x.split(" ")

        num_ = [int(i) for i in num.split("-")]
        char = char[0]
        if num_[0] <= Counter(pwd)[char] <= num_[1]:
            valid += 1

    assert valid == 447


# %% Part 2


def test2() -> None:
    valid = 0
    for x in raw:
        num, char, pwd = x.split(" ")

        n = [int(i) for i in num.split("-")]
        char = char[0]

        if (pwd[n[0] - 1] == char) != (pwd[n[1] - 1] == char):  # XOR
            valid += 1

    assert valid == 249


# %%
